dask/compose.py

```diff
@@ -0,0 +1,299 @@
+            return (type(task), args), dasks
+        else:
+            return args, dasks
+    elif isinstance(task, dict):
+        args, dasks = to_task_dasks(list([k, v] for k, v in task.items()))
+        return (dict, args), dasks
+    else:
+        return task, []
+
+
+def tokenize(v):
+    """Mapping function from task -> consistent name"""
+    # TODO: May have hash collisions...
+    try:
+        return str(hash(v))
+    except TypeError:
+        pass
+    return str(hash(str(v)))
+
+
+def applyfunc(func, *args, **kwargs):
+    """Create a Value by applying a function to args.
+
+    Given a function and arguments, return a Value that represents the result
+    of that computation."""
+
+    args, dasks = unzip(map(to_task_dasks, args), 2)
+    dasks = flat_unique(dasks)
+    name = tokenize((func, args, frozenset(kwargs.items())))
+    if kwargs:
+        func = partial(func, **kwargs)
+    dasks.append({name: (func,) + args})
+    return Value(name, dasks)
+
+
+def do(func):
+    """Wraps a function so that it outputs a ``Value``.
+
+    Examples
+    --------
+    Can be used as a decorator:
+
+    >>> @do
+    ... def add(a, b):
+    ...     return a + b
+    >>> res = add(1, 2)
+    >>> type(res) == Value
+    True
+    >>> res.compute()
+    3
+
+    For other cases, it may be cleaner to call ``do`` on a function at call
+    time:
+
+    >>> res2 = do(sum)([res, 2, 3])
+    >>> res2.compute()
+    8
+    """
+    @wraps(func)
+    def _dfunc(*args, **kwargs):
+        return applyfunc(func, *args, **kwargs)
+    return _dfunc
+
+
+def optimize(dsk, keys):
+    dsk2 = cull(dsk, keys)
+    return fuse(dsk2)
+
+
+def get(dsk, keys, get=None, **kwargs):
+    """Specialized get function"""
+    get = get or _globals['get'] or threaded.get
+    dsk2 = optimize(dsk, keys)
+    return get(dsk2, keys, **kwargs)
+
+
+def right(method):
+    """Wrapper to create 'right' version of operator given left version"""
+    def _inner(self, other):
+        return method(other, self)
+    return _inner
+
+
+class Value(object):
+    """Represents a value to be computed by dask.
+
+    Equivalent to the output from a single key in a dask graph.
+    """
+    __slots__ = ('_key', '_dasks')
+
+    def __init__(self, name, dasks):
+        object.__setattr__(self, '_key', name)
+        object.__setattr__(self, '_dasks', dasks)
+
+    def compute(self, **kwargs):
+        """Compute the result."""
+        dask1 = merge(*self._dasks)
+        dask2 = cull(dask1, self.key)
+        return get(dask2, self.key, **kwargs)
+
+    @property
+    def dask(self):
+        return merge(*self._dasks)
+
+    @property
+    def key(self):
+        return self._key
+
+    def visualize(self, optimize_graph=False, **kwargs):
+        """Visualize the dask as a graph"""
+        from dask.dot import dot_graph
+        if optimize_graph:
+            dot_graph(optimize(self.dask, self.key), **kwargs)
+        else:
+            dot_graph(self.dask, **kwargs)
```

Comment:
As of https://github.com/ContinuumIO/dask/pull/424 these should return the value of `dot_graph`